lib/codecs/src/decoding/format/avro.rs

```diff
@@ -0,0 +1,366 @@
+        Ok(smallvec![event])
+    }
+}
+
+// Can't use std::convert::TryFrom because of orphan rules
+pub fn try_from(value: AvroValue) -> vector_common::Result<VrlValue> {
+    // Very similar to avro to json see `impl std::convert::TryFrom<AvroValue> for serde_json::Value`
+    // LogEvent has native support for bytes, so it is used for Bytes and Fixed
+    match value {
+        AvroValue::Array(array) => {
+            let mut vector = Vec::new();
+            for item in array {
+                vector.push(try_from(item)?);
+            }
+            Ok(VrlValue::Array(vector))
+        }
+        AvroValue::Boolean(boolean) => Ok(VrlValue::from(boolean)),
+        AvroValue::Bytes(bytes) => Ok(VrlValue::from(bytes)),
+        AvroValue::Date(d) => Ok(VrlValue::from(d)),
+        AvroValue::Decimal(ref d) => Ok(<Vec<u8>>::try_from(d)
+            .map(|vec| VrlValue::Array(vec.into_iter().map(VrlValue::from).collect()))?),
+        AvroValue::Double(double) => Ok(VrlValue::from_f64_or_zero(double)),
+        AvroValue::Duration(d) => Ok(VrlValue::Array(
+            <[u8; 12]>::from(d)
+                .into_iter()
+                .map(VrlValue::from)
+                .collect(),
+        )),
+        AvroValue::Enum(_, string) => Ok(VrlValue::from(string)),
+        AvroValue::Fixed(_, bytes) => Ok(VrlValue::from(bytes)),
+        AvroValue::Float(float) => Ok(VrlValue::from_f64_or_zero(float as f64)),
+        AvroValue::Int(int) => Ok(VrlValue::from(int)),
+        AvroValue::Long(long) => Ok(VrlValue::from(long)),
+        AvroValue::Map(items) => items
+            .into_iter()
+            .map(|(key, value)| try_from(value).map(|v| (key, v)))
+            .collect::<Result<Vec<_>, _>>()
+            .map(|v| VrlValue::Object(v.into_iter().collect())),
+        AvroValue::Null => Ok(VrlValue::Null),
+        AvroValue::Record(items) => items
+            .into_iter()
+            .map(|(key, value)| try_from(value).map(|v| (key, v)))
+            .collect::<Result<Vec<_>, _>>()
+            .map(|v| VrlValue::Object(v.into_iter().collect())),
+        AvroValue::String(string) => Ok(VrlValue::from(string)),
+        AvroValue::TimeMicros(time_micros) => Ok(VrlValue::from(time_micros)),
+        AvroValue::TimeMillis(time_millis) => Ok(VrlValue::from(time_millis)),
+        AvroValue::TimestampMicros(micros) => Ok(VrlValue::Timestamp(
+            Utc.timestamp_micros(micros)
+                .single()
+                .ok_or("failed to convert TimestampMicros")?,
+        )),
+        AvroValue::TimestampMillis(millis) => Ok(VrlValue::Timestamp(
+            Utc.timestamp_millis_opt(millis)
+                .single()
+                .ok_or("failed to convert TimestampMillis")?,
+        )),
+        AvroValue::Union(_, v) => try_from(*v),
+        AvroValue::Uuid(uuid) => Ok(VrlValue::from(uuid.as_hyphenated().to_string())),
+        AvroValue::LocalTimestampMillis(_) | AvroValue::LocalTimestampMicros(_) => unimplemented!(),
```

Comment:
Let's return an error here instead.

Reply:
Like TimestampMillis(i64) and TimestampMicros(i64), can we let them as is and let user deal it manually?

Reply:
Correct me if I am wrong, here we will panic when we hit this branch. I don't think we wan't to pursue more complicated solutions here. We should probably just do this:

```
return Err("format!("{} type is not supported", std::any::type_name::<T>())
```

Reply:
Or we can let it as it here?
```diff
diff --git a/lib/codecs/src/decoding/format/avro.rs b/lib/codecs/src/decoding/format/avro.rs
index 3bd6859792..fd0cf90c7e 100644
--- a/lib/codecs/src/decoding/format/avro.rs
+++ b/lib/codecs/src/decoding/format/avro.rs
@@ -211,19 +211,12 @@ pub fn try_from(value: AvroValue) -> vector_common::Result<VrlValue> {
         AvroValue::String(string) => Ok(VrlValue::from(string)),
         AvroValue::TimeMicros(time_micros) => Ok(VrlValue::from(time_micros)),
         AvroValue::TimeMillis(time_millis) => Ok(VrlValue::from(time_millis)),
-        AvroValue::TimestampMicros(micros) => Ok(VrlValue::Timestamp(
-            Utc.timestamp_micros(micros)
-                .single()
-                .ok_or("failed to convert TimestampMicros")?,
-        )),
-        AvroValue::TimestampMillis(millis) => Ok(VrlValue::Timestamp(
-            Utc.timestamp_millis_opt(millis)
-                .single()
-                .ok_or("failed to convert TimestampMillis")?,
-        )),
+        AvroValue::TimestampMicros(micros) => Ok(VrlValue::from(micros)),
+        AvroValue::TimestampMillis(millis) => Ok(VrlValue::from(millis)),
         AvroValue::Union(_, v) => try_from(*v),
         AvroValue::Uuid(uuid) => Ok(VrlValue::from(uuid.as_hyphenated().to_string())),
-        AvroValue::LocalTimestampMillis(_) | AvroValue::LocalTimestampMicros(_) => unimplemented!(),
+        AvroValue::LocalTimestampMillis(millis) => Ok(VrlValue::from(millis)),
+        AvroValue::LocalTimestampMicros(micros) => Ok(VrlValue::from(micros)),
     }
 }
```

Reply:
> Like TimestampMillis(i64) and TimestampMicros(i64), can we let them as is and let user deal it manually?

Using the Utc API makes sense here. I was referring to all other types for which the mapping is not obvious/complex.